test/manual/kerberos.test.js

```diff
@@ -79,18 +80,179 @@ describe('Kerberos', function () {
     });
   });
 
-  it('validate that CANONICALIZE_HOST_NAME can be passed in', function (done) {
-    if (process.platform === 'darwin') {
-      this.test.skipReason = 'DNS does not resolve with proper CNAME record on evergreen MacOS';
-      this.skip();
+  context('when passing in CANONICALIZE_HOST_NAME', function () {
+    beforeEach(function () {
+      //if (process.platform === 'darwin') {
```

Comment:
Looks like this is still WIP - but we probably want to either uncomment these or delete them before we merge.

Also, to summarize discussions from the earlier calls, for both success and failure cases, wherever possible, it would be good to assert the result from the correct function is returned (not just counting function calls, because, among other things, function call counts don't ensure any particular order of calls) or what those functions are called with. We can do that by making sure the functions resolve with different values and then checking that the callback is called with the one from the expected function call.

Reply:
Just scanning I think that the 'when the cname lookup is empty' test section could test the value of the hostname returned, it would be some mock value, but the empty case at least hands back what was passed in. 

Separate point: 
Would it help to move the testing of this to a unit file and export the `performGSSAPICanonicalizeHostName` function? I think we have a good idea of the very explicit order of events that need be tested, and putting it in unit would just mean not needing what ever env variables are needed to spin up the actual int tests.

Reply:
I've uncommented that. With respect to asserting the number of function calls vs what they were called with, it's a bit tricky in this case. Firstly, these tests never stub `dns.lookup` as this is not the only place it is used and the stubs are on the entire module. So these tests are stubbing the reverse ptr lookup and the cname lookups since they are only used in the canonicalization case. With the reverse ptr lookup, that would need to assert on the stub that it was passed an IP address, which I didn't want to hard code in the test and attempting the lookup would be yet another call in the test to `dns.lookup`, which we already had a spy on. (I felt like it was going in circles). :) For the cname assertions I've changed this from `calledOnce` to `calledOnceWith` to ensure the host was being passed in.

Reply:
Could the stub/real method concern be worked around with the sinon callThrough method on stubs, and the conditional onFirstCall/onSecondCall type definitions for what the stub returns? (Like in the last couple of comments in this [thread](https://stackoverflow.com/questions/37343232/sinon-js-only-stub-a-method-once))

There's also Neal's suggestion to make it more of a unit test. If we have to, I don't think we absolutely must hold up the feature on this, but at the same time I am hesitant to have a feature where we don't actually fully test that the intended behavior is executed for each of the specified options.

Reply:
My preference would be to create a new ticket as a follow up for testing, and separate these into unit tests as @nbbeeken suggested by exporting the `performGSSAPICanonicalizeHostName` function. Then when the BUILD ticket is done for the DNS changes I can update the auth spec with prose tests for Kerberos and ensure our manual tests follow the prose tests.